watcher/watcher.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watcher
+
+import (
+	"context"
+
+	"perun.network/go-perun/channel"
+)
+
+type (
+	// Watcher is the interface used by client to interact with the Watcher.
+	//
+	// When a new channel is established, the client should register it with
+	// the watcher by calling the StartWatching method.
+	//
+	// After that, it could publish each state on the StatesPub. If any state
+	// is registered or progressed on the blockchain, the corresponding
+	// adjudicator event will be relayed by the watcher on the AdjudicatorSub.
+	// If the registered state is not the latest state (published to the
+	// watcher) then, the watcher will automatically refute by registering the
+	// latest state on the blockchain.
+	//
+	// Once a channel is closed, client can de-register the channel from the
+	// watcher by calling the StopWatching API. This will close the
+	// AdjudicatorSub and StatesSub. Client must not publish states after
+	// this.
+	Watcher interface {
+		StartWatchingLedgerChannel(context.Context, channel.SignedState) (StatesPub, AdjudicatorSub, error)
+		StartWatchingSubChannel(context.Context, channel.ID, channel.SignedState) (StatesPub, AdjudicatorSub, error)
+		StopWatching(channel.ID) error
```

Comment:
Yes, I agree.

Though, we might not use the context in the local watcher per-say, this will become for remote watcher implementations.

Reply:
I had added a context.

Since the review is approved, I have squashed all the review commits. See the [diff](https://github.com/hyperledger-labs/go-perun/compare/22c7d1a99fe23c3b7236ae325a6a5e1554e2eb0d..7c6a0fa04c3f6bc6ba571081d7b825f9930849ca) of this force-push.